File: C-Sharp/Source/Virtusales.Biblio.API.Session.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Net;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Json;

namespace Virtusales.Biblio.API
{
	internal class Session : BiblioAPIBase
	{
		private RequestObject<LoginRequestData, LoginResponseData> LoginRequest;
		private string Username;
		private string Password;

		public string ErrorMessage = "";

		private string _Token;
		public string Token {
			get { return _Token; }
			private set { _Token = value; }
		}

		public Session(string applicationName, string url, string username, string password)
		{
			this.Username = username;
			this.Password = password;
			this.LoginRequest = new RequestObject<LoginRequestData, LoginResponseData>(applicationName, url, "APILogin");
			this.LoginRequest.DebugToConsole = this.DebugToConsole;
		}

		public bool Login(ref string errorMessage = "")
		{
			LoginRequestData RequestData = new LoginRequestData {
				Username = Username,
				Password = Password
			};
			ResponseObject<LoginResponseData> Response = new ResponseObject<LoginResponseData>();

			Token = "";

			Response = LoginRequest.PostRequest(RequestData);

			if (Response.Success) {
				if (Response.Value.Success) {
					Token = Response.Value.SessionToken;
				} else {
					errorMessage = "Reason: " + Response.Value.Detail;
					return false;
				}
			} else if (Response.StatusCode == HttpStatusCode.Unauthorized) {
				errorMessage = "Unauthorised: " + Response.ErrorMessage;
				return false;
			} else {
				errorMessage = Response.ErrorMessage + Constants.vbCrLf + "Response HTTP State Code was: " + Convert.ToInt32(Response.StatusCode) + " " + Response.StatusCode.ToString() + Constants.vbCrLf + "Response was: " + Response.RawText;
				return false;
			}

			return true;
		}

		[DataContract()]
		private class LoginRequestData
		{
			[DataMember(Name = "username", IsRequired = true)]
			public string Username;
			[DataMember(Name = "password", IsRequired = true)]
			public string Password;
		}

		[DataContract()]
		private class LoginResponseData
		{
			[DataMember(Name = "success", IsRequired = true)]
			public bool Success;
			[DataMember(Name = "detail", IsRequired = true)]
			public string Detail;
			[DataMember(Name = "sessiontoken", IsRequired = true)]
			public string SessionToken;
		}
	}
}

File: VBnet/Source/Virtusales.Biblio.API.Session.vb
Imports System.Net
Imports System.Runtime.Serialization
Imports System.Runtime.Serialization.Json

Namespace Virtusales.Biblio.API
    Friend Class Session : Inherits BiblioAPIBase
        Private LoginRequest As RequestObject(Of LoginRequestData, LoginResponseData)
        Private Username As String
        Private Password As String
        
        Public ErrorMessage As String = ""
        
        Private _Token As String
        Public Property Token As String
            Get
                Return _Token
            End Get
            Private Set(Value As String)
                _Token = Value
            End Set
        End Property
    
        Public Sub New(applicationName As String, url As String, username As String, password As String)
            Me.Username = username
            Me.Password = password
            Me.LoginRequest = New RequestObject(Of LoginRequestData, LoginResponseData)(applicationName, url, "APILogin")
            Me.LoginRequest.DebugToConsole = Me.DebugToConsole
        End Sub  
    
        Public Function Login(ByRef Optional errorMessage As String = "") As Boolean
            Dim RequestData As New LoginRequestData With { .Username = Username, .Password = Password }
            Dim Response As New ResponseObject(Of LoginResponseData)
            
            Token = ""
            
            Response = LoginRequest.PostRequest(RequestData)
            
            If Response.Success Then
                If Response.Value.Success Then
                    Token = Response.Value.SessionToken
                Else
                    ErrorMessage = "Reason: " & Response.Value.Detail
                    Return False
                End If
            Else If Response.StatusCode = HttpStatusCode.Unauthorized Then
                errorMessage = "Unauthorised: " & Response.ErrorMessage
                Return False
            Else
                errorMessage =  Response.ErrorMessage & vbCrLf & _
                                "Response HTTP State Code was: " & CInt(Response.StatusCode) & " " & Response.StatusCode.ToString() & vbCrLf & _
                                "Response was: " & Response.RawText
                Return False
            End If
            
            Return True
        End Function
    
        <DataContract>
        Private Class LoginRequestData
            <DataMember(Name := "username", IsRequired := True)> Public Username As String
            <DataMember(Name := "password", IsRequired := True)> Public Password As String
        End Class    
        
        <DataContract>
        Private Class LoginResponseData
            <DataMember(Name := "success", IsRequired := True)> Public Success As Boolean
            <DataMember(Name := "detail", IsRequired := True)> Public Detail As String
            <DataMember(Name := "sessiontoken", IsRequired := True)> Public SessionToken As String
        End Class  
    End Class
End Namespace